# (3주차) 9월28일
> 파이토치를 이용하여 회귀모형 학습하기 (2) 

- toc:true
- branch: master
- badges: true
- comments: true
- author: 최규빈

---

### Import 

In [1]:
import torch 
import numpy as np 

### Data

`-` model: $y_i= w_0+w_1 x_i +\epsilon_i = 2.5 + 4x_i +\epsilon_i, \quad i=1,2,\dots,n$ 

`-` model: ${\bf y}={\bf X}{\bf W} +\boldsymbol{\epsilon}$

- ${\bf y}=\begin{bmatrix} y_1 \\ y_2 \\ \dots \\ y_n\end{bmatrix}, \quad {\bf X}=\begin{bmatrix} 1 & x_1 \\ 1 & x_2 \\ \dots \\ 1 & x_n\end{bmatrix}, \quad {\bf W}=\begin{bmatrix} 2.5 \\ 4 \end{bmatrix}, \quad \boldsymbol{\epsilon}= \begin{bmatrix} \epsilon_1 \\ \dots \\ \epsilon_n\end{bmatrix}$

In [2]:
torch.manual_seed(43052)
n=100
ones= torch.ones(n)
x,_ = torch.randn(n).sort()
X = torch.vstack([ones,x]).T
W = torch.tensor([2.5,4])
ϵ = torch.randn(n)*0.5
y = X@W + ϵ
ytrue = X@W

### 이전방법요약 

`-` step1: yhat 

`-` step2: loss 

`-` step3: derivation 

`-` step4: update 

---

### step1: yhat

`-` feedforward 신경망을 설계하는 과정

`-` 이 단계가 잘 완료되었다면, 임의의 ${\bf\hat{W}}$을 넣었을 때 $\bf\hat{y}$를 계산할 수 있어야 함 

#### 방법1: 직접선언 (내가 공식을 알고 있어야 한다)

In [6]:
What=torch.tensor([-5.0,10.0],requires_grad=True)  #미분 꼬리표=requires_grad=True 기억하기!

In [7]:
yhat1=X@What

In [8]:
yhat1

tensor([-29.8211, -28.6215, -24.9730, -21.2394, -19.7919, -19.6354, -19.5093,
        -19.4352, -18.7223, -18.0793, -16.9040, -16.0918, -16.0536, -15.8746,
        -14.4690, -14.3193, -13.6426, -12.8578, -12.5486, -12.4213, -11.9484,
        -11.1034, -10.8296, -10.6210, -10.5064, -10.0578,  -9.8063,  -9.7380,
         -9.7097,  -9.6756,  -8.8736,  -8.7195,  -8.6880,  -8.1592,  -7.7752,
         -7.7716,  -7.7339,  -7.7208,  -7.6677,  -7.1551,  -7.0004,  -6.8163,
         -6.7081,  -6.5655,  -6.4480,  -6.3612,  -6.0566,  -5.6031,  -5.5589,
         -5.2137,  -4.3446,  -4.3165,  -3.8047,  -3.5801,  -3.4793,  -3.4325,
         -2.3545,  -2.3440,  -1.8434,  -1.7799,  -1.5386,  -1.0161,  -0.8103,
          0.4426,   0.5794,   0.9125,   1.1483,   1.4687,   1.4690,   1.5234,
          1.6738,   2.0592,   2.1414,   2.8221,   3.1536,   3.6682,   4.2907,
          4.8037,   4.8531,   4.9414,   5.3757,   5.3926,   5.6973,   6.0239,
          6.1261,   6.5317,   7.2891,   8.4032,   8.4936,   9.27

#### (★) 방법2: torch.nn.Linear() 사용
    - nn안에 linear라는 클래스가 있음

In [9]:
net = torch.nn.Linear(in_features=2 ,out_features=1, bias=False) 

In [10]:
net.weight.data

tensor([[0.3320, 0.1982]])

In [11]:
net.weight.data=torch.tensor([[-5.0,10.0]])

In [12]:
net.weight.data

tensor([[-5., 10.]])

In [13]:
net(X)

tensor([[-29.8211],
        [-28.6215],
        [-24.9730],
        [-21.2394],
        [-19.7919],
        [-19.6354],
        [-19.5093],
        [-19.4352],
        [-18.7223],
        [-18.0793],
        [-16.9040],
        [-16.0918],
        [-16.0536],
        [-15.8746],
        [-14.4690],
        [-14.3193],
        [-13.6426],
        [-12.8578],
        [-12.5486],
        [-12.4213],
        [-11.9484],
        [-11.1034],
        [-10.8296],
        [-10.6210],
        [-10.5064],
        [-10.0578],
        [ -9.8063],
        [ -9.7380],
        [ -9.7097],
        [ -9.6756],
        [ -8.8736],
        [ -8.7195],
        [ -8.6880],
        [ -8.1592],
        [ -7.7752],
        [ -7.7716],
        [ -7.7339],
        [ -7.7208],
        [ -7.6677],
        [ -7.1551],
        [ -7.0004],
        [ -6.8163],
        [ -6.7081],
        [ -6.5655],
        [ -6.4480],
        [ -6.3612],
        [ -6.0566],
        [ -5.6031],
        [ -5.5589],
        [ -5.2137],


In [14]:
yhat2=net(X)

#### 방법3: torch.nn.Linear()사용, bias=True

In [15]:
net = torch.nn.Linear(in_features=1 ,out_features=1, bias=True) 

`-` 입력차원을 1로 했기 때문에 net.weight.data 값이 1개만 나온다  
`-` 또 bias를 전과 다르게 True로 줘서 아래 bias.data도 가능하다

In [16]:
net.weight.data

tensor([[0.3480]])

In [18]:
net.weight.data=torch.tensor([[10.0]])

In [19]:
net.bias.data=torch.tensor([-5.0])

In [20]:
net.weight,net.bias

(Parameter containing:
 tensor([[10.]], requires_grad=True),
 Parameter containing:
 tensor([-5.], requires_grad=True))

In [23]:
net(x)   #차원오류

RuntimeError: mat1 and mat2 shapes cannot be multiplied (1x100 and 1x1)

In [24]:
net(x.reshape(100,1))   #shape을 바꿔준다

tensor([[-29.8211],
        [-28.6215],
        [-24.9730],
        [-21.2394],
        [-19.7919],
        [-19.6354],
        [-19.5093],
        [-19.4352],
        [-18.7223],
        [-18.0793],
        [-16.9040],
        [-16.0918],
        [-16.0536],
        [-15.8746],
        [-14.4690],
        [-14.3193],
        [-13.6426],
        [-12.8578],
        [-12.5486],
        [-12.4213],
        [-11.9484],
        [-11.1034],
        [-10.8296],
        [-10.6210],
        [-10.5064],
        [-10.0578],
        [ -9.8063],
        [ -9.7380],
        [ -9.7097],
        [ -9.6756],
        [ -8.8736],
        [ -8.7195],
        [ -8.6880],
        [ -8.1592],
        [ -7.7752],
        [ -7.7716],
        [ -7.7339],
        [ -7.7208],
        [ -7.6677],
        [ -7.1551],
        [ -7.0004],
        [ -6.8163],
        [ -6.7081],
        [ -6.5655],
        [ -6.4480],
        [ -6.3612],
        [ -6.0566],
        [ -5.6031],
        [ -5.5589],
        [ -5.2137],


#### 

### step2: loss

#### 방법1: 손실함수를 직접정의하는 방법

In [28]:
loss=torch.mean((y-yhat1)**2)
loss

tensor(85.8769, grad_fn=<MeanBackward0>)

In [26]:
loss=torch.mean((y-yhat2)**2)
loss

tensor(176.2661, grad_fn=<MeanBackward0>)

- 왜 다르지?

In [30]:
y-yhat2

tensor([[ 21.2791,  23.2444,  23.8716,  ...,  42.9194,  42.3679,  43.6551],
        [ 20.0794,  22.0448,  22.6719,  ...,  41.7197,  41.1683,  42.4555],
        [ 16.4309,  18.3962,  19.0234,  ...,  38.0712,  37.5197,  38.8070],
        ...,
        [-29.5981, -27.6328, -27.0056,  ...,  -7.9578,  -8.5093,  -7.2220],
        [-29.5986, -27.6333, -27.0062,  ...,  -7.9583,  -8.5098,  -7.2226],
        [-30.1744, -28.2091, -27.5820,  ...,  -8.5341,  -9.0856,  -7.7984]],
       grad_fn=<SubBackward0>)

In [31]:
(y-yhat2).shape

torch.Size([100, 100])

- `y`는 길이가 100인 벡터, `yhat2`는 100X1 matrix임
- 2개가 계산되면서 생긴 오류임
- 176.2661? 이건 잘못된 결과임

In [34]:
#해결방법 : flatten으로 벡터화하기
torch.mean((y-yhat2.flatten())**2)

tensor(85.8769, grad_fn=<MeanBackward0>)

In [35]:
#해결방법 : y를 reshpae 해주기
loss=torch.mean((y.reshape(100,1)-yhat2)**2)
loss

tensor(85.8769, grad_fn=<MeanBackward0>)

#### 방법2: torch.nn.MSELoss()를 사용하여 손실함수를 정의하는 방법 

In [36]:
lossfn=torch.nn.MSELoss()

In [37]:
loss=lossfn(y,yhat1)
loss

tensor(85.8769, grad_fn=<MseLossBackward>)

In [38]:
loss=lossfn(y.reshape(100,1),yhat2)
loss

tensor(85.8769, grad_fn=<MseLossBackward>)

### [숙제](https://ieilms.jbnu.ac.kr/)

`-` model: $y_i= w_0+w_1 x_{i1}+w_2 x_{i2} +\epsilon_i = 2.5 + 4x_{1i} + -2x_{2i}+\epsilon_i, \quad i=1,2,\dots,n$ 

In [40]:
torch.manual_seed(43052)
n=100
ones= torch.ones(n)
x1,_ = torch.randn(n).sort()
x2,_ = torch.randn(n).sort()
X = torch.vstack([ones,x1,x2]).T
W = torch.tensor([2.5,4,-2])
ϵ = torch.randn(n)*0.5
y = X@W + ϵ
ytrue = X@W

In [41]:
X

tensor([[ 1.0000, -2.4821, -2.3721],
        [ 1.0000, -2.3621, -2.3032],
        [ 1.0000, -1.9973, -2.2271],
        [ 1.0000, -1.6239, -2.0301],
        [ 1.0000, -1.4792, -1.9157],
        [ 1.0000, -1.4635, -1.8241],
        [ 1.0000, -1.4509, -1.6696],
        [ 1.0000, -1.4435, -1.6675],
        [ 1.0000, -1.3722, -1.4723],
        [ 1.0000, -1.3079, -1.4405],
        [ 1.0000, -1.1904, -1.4111],
        [ 1.0000, -1.1092, -1.3820],
        [ 1.0000, -1.1054, -1.3803],
        [ 1.0000, -1.0875, -1.3456],
        [ 1.0000, -0.9469, -1.3255],
        [ 1.0000, -0.9319, -1.2860],
        [ 1.0000, -0.8643, -1.2504],
        [ 1.0000, -0.7858, -1.2095],
        [ 1.0000, -0.7549, -1.1498],
        [ 1.0000, -0.7421, -1.1151],
        [ 1.0000, -0.6948, -1.0980],
        [ 1.0000, -0.6103, -1.0609],
        [ 1.0000, -0.5830, -0.9825],
        [ 1.0000, -0.5621, -0.9672],
        [ 1.0000, -0.5506, -0.9396],
        [ 1.0000, -0.5058, -0.9208],
        [ 1.0000, -0.4806, -0.8768],
 

`-` torch.nn.Linear() 를 이용하여 $\bf{\hat{W}}=\begin{bmatrix}1 \\ 1 \\ 1 \end{bmatrix}$ 에 대한 $\hat{y}$를 구하라. 

In [49]:
net = torch.nn.Linear(in_features=3,out_features=1, bias=False) 

In [50]:
net.weight.data

tensor([[ 0.0411,  0.3420, -0.5768]])

In [51]:
net.weight.data=torch.tensor([[1.0,1.0,1.0]])

In [52]:
net.weight.data

tensor([[1., 1., 1.]])

In [63]:
net(X)

tensor([[-3.8542],
        [-3.6654],
        [-3.2244],
        [-2.6540],
        [-2.3949],
        [-2.2877],
        [-2.1205],
        [-2.1110],
        [-1.8446],
        [-1.7484],
        [-1.6015],
        [-1.4912],
        [-1.4857],
        [-1.4330],
        [-1.2724],
        [-1.2179],
        [-1.1147],
        [-0.9953],
        [-0.9047],
        [-0.8572],
        [-0.7928],
        [-0.6712],
        [-0.5655],
        [-0.5293],
        [-0.4903],
        [-0.4266],
        [-0.3574],
        [-0.2255],
        [-0.1800],
        [-0.1702],
        [-0.0791],
        [-0.0280],
        [ 0.0160],
        [ 0.1480],
        [ 0.2441],
        [ 0.3293],
        [ 0.3503],
        [ 0.3997],
        [ 0.4105],
        [ 0.4984],
        [ 0.5157],
        [ 0.5393],
        [ 0.5820],
        [ 0.6235],
        [ 0.6382],
        [ 0.6687],
        [ 0.7057],
        [ 0.7568],
        [ 0.7995],
        [ 0.9063],
        [ 0.9989],
        [ 1.0058],
        [ 1.